public/app/features/alerting/unified/components/settings/SettingsContext.tsx

```diff
@@ -0,0 +1,181 @@
+import { union, without, debounce } from 'lodash';
+import React, { PropsWithChildren, useEffect, useRef } from 'react';
+
+import { AppEvents } from '@grafana/data';
+import { getAppEvents } from '@grafana/runtime';
+import { AlertmanagerChoice, ExternalAlertmanagerConfig } from 'app/plugins/datasource/alertmanager/types';
+import { dispatch } from 'app/store/store';
+
+import { alertmanagerApi } from '../../api/alertmanagerApi';
+import { dataSourcesApi, enableOrDisableHandlingGrafanaManagedAlerts } from '../../api/dataSourcesApi';
+import {
+  ExternalAlertmanagerDataSourceWithStatus,
+  useExternalDataSourceAlertmanagers,
+} from '../../hooks/useExternalAmSelector';
+import { deleteAlertManagerConfigAction, updateAlertManagerConfigAction } from '../../state/actions';
+import { GRAFANA_RULES_SOURCE_NAME, isAlertmanagerDataSourceInterestedInAlerts } from '../../utils/datasource';
+
+const appEvents = getAppEvents();
+const USING_INTERNAL_ALERTMANAGER_SETTINGS = [AlertmanagerChoice.Internal, AlertmanagerChoice.All];
+
+interface Context {
+  deliverySettings?: ExternalAlertmanagerConfig;
+  externalAlertmanagerDataSourcesWithStatus: ExternalAlertmanagerDataSourceWithStatus[];
+
+  isLoading: boolean;
+  isUpdating: boolean;
+
+  // for enabling / disabling Alertmanager datasources as additional receivers
+  enableAlertmanager: (uid: string) => void;
+  disableAlertmanager: (uid: string) => void;
+
+  // for updating or resetting the configuration for an Alertmanager
+  updateAlertmanagerSettings: (name: string, oldConfig: string, newConfig: string) => void;
+  resetAlertmanagerSettings: (name: string) => void;
+}
+
+const SettingsContext = React.createContext<Context | undefined>(undefined);
+const isInternalAlertmanager = (uid: string) => uid === GRAFANA_RULES_SOURCE_NAME;
+
+export const SettingsProvider = (props: PropsWithChildren) => {
+  // this list will keep track of Alertmanager UIDs (including internal) that are interested in receiving alert instances
+  // this will be used to infer the correct "delivery mode" and update the correct list of datasources with "wantsAlertsReceived"
+  let interestedAlertmanagers: string[] = [];
+
+  const { currentData: deliverySettings, isLoading: isLoadingDeliverySettings } =
+    alertmanagerApi.endpoints.getExternalAlertmanagerConfig.useQuery();
+
+  const [updateDeliverySettings, updateDeliverySettingsState] =
+    alertmanagerApi.endpoints.saveExternalAlertmanagersConfig.useMutation();
+  const [enableGrafanaManagedAlerts, enableOrDisableHandlingGrafanaManagedAlertsState] =
+    enableOrDisableHandlingGrafanaManagedAlerts();
+
+  const externalAlertmanagersWithStatus = useExternalDataSourceAlertmanagers();
+
+  const interestedInternal = USING_INTERNAL_ALERTMANAGER_SETTINGS.some(
```

Comment:
This looks very similar to `isReceivingOnInternalAlertmanager`
Could we extract util functions for that type of checking?